cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala

```diff
@@ -128,157 +218,227 @@ class RandomBoundedTraverser(graph: Graph, dir: GraphDir, homeNodeIds: Seq[Int],
     walkParams.filterHomeNodeByNumEdges) with BoundedIterator[Node]
 
 /**
- * A traverser that keeps track of distance from homeNodeIds. If {@code onlyOnce} is true,
- * the same node is not traversed again.
+ * Trait defining an object that keeps track of seen nodes.
  */
-abstract class DistanceTraverser[T](graph: Graph, homeNodeIds: Seq[Int], onlyOnce: Boolean)
-    extends Traverser { self =>
+trait NodesSeenBefore {
+  protected def seenBefore(id: Int): Boolean
+}
 
-  private val distanceTracker = new IntInfoKeeper(self.onlyOnce)
+trait MaxDepthLimit {
+  val maxDepth: Option[Int]
+  def maxDepthReached(depth: Int): Boolean = {
+    maxDepth match {
+      case Some(d) if depth >= d => true
+      case _ => false
+    }
+  }
+}
 
-  protected def init(defaultInfo: T) {
-    homeNodeIds foreach { id => visitPotentialNode(id, defaultInfo, 0) }
+trait DegreeLimit {
+  val maxNumEdgesThreshold: Option[Int]
+  def degreeToHigh(degree: Int): Boolean = {
+    maxNumEdgesThreshold match {
+      case Some(d) if degree > d => true
+      case _ => false
+    }
   }
+}
 
-  protected def visitPotentialNode(id: Int, info: T, dist: Int) {
-    val seen = seenBefore(id)
-    if (!(onlyOnce && seen)) {
-      storeForVisit(id, info)
+/**
+ * A traverser that keeps track of visiting distance from `homeNodeIds`.
+ */
+trait PathLengthTracker extends QueueBasedTraverser with NodesSeenBefore {
+  private lazy val nextVisitDistanceTracker = new IntInfoKeeper(false)
+  private lazy val visitDistanceTracker = new IntInfoKeeper(true)
+
+  override protected def enqueue(nodes: Seq[Int], from: Option[Int]): Unit = {
+    val fromDistance = from.flatMap(visitDistanceTracker.infoOfNode).getOrElse(-1)
+    nodes foreach { id =>
+      nextVisitDistanceTracker.recordInfo(id, fromDistance + 1)
     }
-    if (!seen) distanceTracker.recordInfo(id, dist)
+    super.enqueue(nodes, from)
   }
 
-  protected def storeForVisit(id: Int, info: T)
+  override def visitNode(node: Node) {
+    visitDistanceTracker.recordInfo(node.id, nextVisitDistanceTracker.infoOfNode(node).get)
+  }
+
+  override protected def seenBefore(id: Int) = nextVisitDistanceTracker.infoOfNode(id).isDefined
+
+  def distance(id: Int) = visitDistanceTracker.infoOfNode(id)
+}
+
+/**
+ * A traverser that keeps track of first depth of visiting a given node.
+ */
+trait DepthTracker extends QueueBasedTraverser with NodesSeenBefore {
+
+  private lazy val depthTracker = new IntInfoKeeper(true)
+
+  override protected def enqueue(nodes: Seq[Int], from: Option[Int]): Unit = {
+    val fromDepth = from.flatMap(depthTracker.infoOfNode).getOrElse(-1)
+    nodes foreach { id =>
+      depthTracker.recordInfo(id, fromDepth + 1)
+    }
+    super.enqueue(nodes, from)
```

Comment:
IIUC, depth and enqueue time are not the same.
For tree like:
  1
 /  \
2   3
    / 
   4
depth(4) = 2
enqueueTime(3) // so many nodes enqueued before